File: 1866-restore-the-array-from-adjacent-pairs/restore-the-array-from-adjacent-pairs.js
/**
 * @param {number[][]} adjacentPairs
 * @return {number[]}
 */
var restoreArray = function(adjacentPairs) {
    const map = new Map();

    for(let pair of adjacentPairs) {
        if(!map.has(pair[0])) map.set(pair[0], []);
        map.get(pair[0]).push(pair[1]);
        if(!map.has(pair[1])) map.set(pair[1], []);
        map.get(pair[1]).push(pair[0]);
    }

    let start = 0;

    for(let [key, value] of map) {
        if(value.length == 1) {
            start = key;
            break;
        }
    }

    const vis = new Array(adjacentPairs.length+1).fill(false);
    const ans = [];

    function dfs(node) {
        vis[node] = true;

        ans.push(node);

        for(let neigh of map.get(node)) {
            if(!vis[neigh]) {
                dfs(neigh);
            }
        }
    }

    dfs(start);

    return ans;
};

File: 1866-restore-the-array-from-adjacent-pairs/restore-the-array-from-adjacent-pairs.java
class Solution {
    public int[] restoreArray(int[][] adjacentPairs) {
        var map = new HashMap<Integer, Set<Integer>>();
        int start = 0;
        for(int[] pair : adjacentPairs) {
            map.computeIfAbsent(pair[0], a -> new HashSet<>()).add(pair[1]);
            map.computeIfAbsent(pair[1], a -> new HashSet<>()).add(pair[0]);
        }
        for(int key : map.keySet()) {
            if(map.get(key).size() == 1) start = key;
        }
        int[] arr = new int[adjacentPairs.length+1];
        for(int i=0; i<arr.length; i++) {
            arr[i] = start;
            if(i == arr.length-1) return arr;
            var it = map.get(start).iterator();
            //System.out.println(map.get(start));
            int next = it.next();
            if(next == start) next = it.next();
            map.get(start).remove(next);
            map.get(next).remove(start);
            if(map.get(start).size() == 0) map.remove(start);
            start = next;
        }
        return arr;
    }
}